x25519/x25519_test.go

```diff
@@ -0,0 +1,569 @@
+	for i := 0; i < b.N; i++ {
+		if _, _, err := GenerateKey(rand.Reader); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func BenchmarkSignVerify(b *testing.B) {
+	pub, priv, err := GenerateKey(rand.Reader)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	data := make([]byte, 1024)
+	if _, err := io.ReadFull(rand.Reader, data); err != nil {
+		b.Fatal(err)
+	}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		sig, err := Sign(rand.Reader, priv, data)
+		if err != nil {
+			b.Fatal(err)
+		}
+		if !Verify(pub, data, sig) {
+			b.Fatalf("failed to verify: %d", i)
+		}
+	}
+}
+
+func BenchmarkSignVerifyMethod(b *testing.B) {
+	pub, priv, err := GenerateKey(rand.Reader)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	data := make([]byte, 1024)
+	if _, err := io.ReadFull(rand.Reader, data); err != nil {
+		b.Fatal(err)
+	}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		sig, err := priv.Sign(rand.Reader, data, crypto.Hash(0))
+		if err != nil {
+			b.Fatal(err)
+		}
+		if !Verify(pub, data, sig) {
+			b.Fatalf("failed to verify: %d", i)
+		}
+	}
+}
+
+func TestVectors(t *testing.T) {
+	buf := mustTeeReader(t)
+	buf.Write(make([]byte, 64))
+	buf.Write([]byte{
+		0x4e, 0x6c, 0xd9, 0xe8, 0x40, 0x4d, 0xe1, 0xe0,
+		0xc5, 0x22, 0x72, 0x8f, 0x78, 0xde, 0x88, 0x26,
+		0x58, 0x4b, 0x44, 0xef, 0xbc, 0xf6, 0xac, 0x10,
+		0x97, 0x67, 0x60, 0xcc, 0x41, 0xf1, 0x82, 0xb,
+		0xb8, 0x69, 0x6, 0xf3, 0xad, 0xa5, 0x46, 0x6f,
+		0xc2, 0xf3, 0x9, 0x78, 0x7b, 0x3c, 0x6f, 0xd6,
+		0x39, 0x19, 0xb1, 0x48, 0xd6, 0x6, 0xb7, 0x79,
+		0xfb, 0xf7, 0xda, 0x61, 0xee, 0x57, 0x7b, 0xe6,
+	})
+	buf.Write([]byte{
+		0x56, 0x53, 0xd8, 0x94, 0x41, 0x20, 0x4d, 0x9a,
+		0xd6, 0x3c, 0xfa, 0x3a, 0x78, 0x4a, 0x9e, 0x2e,
+		0x4d, 0x13, 0x07, 0xa1, 0x66, 0xde, 0xf7, 0xa7,
+		0xb4, 0x3a, 0xc8, 0x4f, 0x23, 0x2c, 0x9a, 0x6d,
+		0x16, 0x17, 0x10, 0xe6, 0xe7, 0x91, 0x2e, 0x38,
+		0x61, 0xe3, 0xe9, 0x9f, 0x90, 0x2c, 0x6d, 0x9,
+		0x82, 0xc6, 0x44, 0xd6, 0xeb, 0x12, 0xd0, 0x53,
+		0x66, 0x27, 0x61, 0xd0, 0x35, 0x33, 0x31, 0x9d,
+	})
+
+	type args struct {
+		rand    io.Reader
+		message []byte
+		opts    crypto.SignerOpts
+	}
+	tests := []struct {
+		name          string
+		p             PrivateKey
+		args          args
+		wantSignature []byte
+		wantErr       bool
+	}{
+		{"ok", []byte{
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0xbd, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40,
+		}, args{buf, make([]byte, 200), crypto.Hash(0)}, []byte{
+			0x11, 0xc7, 0xf3, 0xe6, 0xc4, 0xdf, 0x9e, 0x8a,
+			0x51, 0x50, 0xe1, 0xdb, 0x3b, 0x30, 0xf9, 0x2d,
+			0xe3, 0xa3, 0xb3, 0xaa, 0x43, 0x86, 0x56, 0x54,
+			0x5f, 0xa7, 0x39, 0x0f, 0x4b, 0xcc, 0x7b, 0xb2,
+			0x6c, 0x43, 0x1d, 0x9e, 0x90, 0x64, 0x3e, 0x4f,
+			0x0e, 0xaa, 0x0e, 0x9c, 0x55, 0x77, 0x66, 0xfa,
+			0x69, 0xad, 0xa5, 0x76, 0xd6, 0x3d, 0xca, 0xf2,
+			0xac, 0x32, 0x6c, 0x11, 0xd0, 0xb9, 0x77, 0x02,
+		}, false},
+		{"ok", []byte{
```

Comment:
Some additional info about the test case in the name may be useful.

Reply:
I added references to the test vectors as comments. All those have to be ok.

Reply:
Great 👍